zokrates_stdlib/stdlib/ecc/edwardsOrderCheck.code

```diff
@@ -0,0 +1,19 @@
+import "ecc/edwardsAdd.code" as add
+import "ecc/edwardsScalarMult.code" as multiply
+import "utils/pack/unpack256.code" as unpack256
+// Verifies that the point is not one of the low-order points.
+// If any of the points is multiplied by the cofactor, the resulting point
+// will be infinity. 
+// Returns 1 if the point is not one of the low-order points, 0 otherwise.
+// Curve parameters are defined with the last argument
+// https://github.com/zcash-hackworks/sapling-crypto/blob/master/src/jubjub/edwards.rs#L166
+def main(field[2] pt, field[10] context) -> (field):
+
+    field cofactor = context[7] 
+    field[256] cofactorExponent = unpack256(cofactor) 
```

Comment:
Making the cofactor this generic costs a lot. Cofactors are usually small, and so multiplying by 4 and 8 would save many constraints.

Reply:
Agreed, in an ideal world the cofactor would take advantage of constant reduction.

For Twisted Edwards curves the cofactor will be `2^c`, so a chain of doubling operations (say, three of them) should be sufficient for a cofactor of 8.

Reply:
Yes, as we will only support Edwards curves we can easily go for this optimisation and save these constraints.

@Schaeff ideally this could also be caught by the optimiser as the cofactor is a constant and all additional iterations in the `scalarMul` loop only create unused variables (see https://github.com/Zokrates/ZoKrates/issues/303)

Reply:
@Schaeff  for reference:

> Unnpack&scalarMul, point hardcoded = 14061 constraints
> Unnpack&scalarMul, point as argument = 14096 constraints
> Double double double, point as argument = 72 constraints

Reply:
We will hardcode the check for now using the double^3 approach, but will leave a not there to make it generic as soon as the optimiser catches these unused variables / assignments.